components/tidb_query/src/rpn_expr/impl_json.rs

```diff
@@ -17,6 +17,75 @@ fn json_type(arg: &Option<Json>) -> Result<Option<Bytes>> {
         .map(|json_arg| Bytes::from(json_arg.json_type())))
 }
 
+#[rpn_fn(raw_varg, min_args=2, extra_validator = json_modify_validator)]
+#[inline]
+fn json_set(args: &[ScalarValueRef]) -> Result<Option<Json>> {
+    json_modify(args, ModifyType::Set)
+}
+
+#[rpn_fn(raw_varg, min_args=2, extra_validator = json_modify_validator)]
+#[inline]
+fn json_insert(args: &[ScalarValueRef]) -> Result<Option<Json>> {
+    json_modify(args, ModifyType::Insert)
+}
+
+#[rpn_fn(raw_varg, min_args=2, extra_validator = json_modify_validator)]
+#[inline]
+fn json_replace(args: &[ScalarValueRef]) -> Result<Option<Json>> {
+    json_modify(args, ModifyType::Replace)
+}
+
+#[inline]
+fn json_modify(args: &[ScalarValueRef], mt: ModifyType) -> Result<Option<Json>> {
+    // args >= 2
+    // base Json argument
+    let base: &Option<Json> = args[0].as_ref();
+    let mut base = base.as_ref().map_or(Json::None, |json| json.to_owned());
+
+    // args.len() / 2
+    let sz = args.len() / 2;
+
+    let mut path_expr_list = Vec::with_capacity(sz);
+    let mut values = Vec::with_capacity(sz);
+
+    for chunk in args[1..].chunks(2) {
+        let path: &Option<Bytes> = chunk[0].as_ref();
+        let value: &Option<Json> = chunk[1].as_ref();
+
+        let json_path = match path.as_ref() {
```

Comment:
can this `match` be integrated to `parse_json_path_expr` as well?

Reply:
> can this `match` be integrated to `parse_json_path_expr` as well?

`parse_json_path_expr` is used in `builtin_json.rs` and other places, maybe I should encapsulate this match and `parse_json_path_expr` in `mod.rs` or in other place?

Reply:
It's up to you to decide. I only found that this `match xxx` seems to be repeated many times even after you extracted a function so I guess that the function can cover this logic as well.

Reply:
> It's up to you to decide. I only found that this `match xxx` seems to be repeated many times even after you extracted a function so I guess that the function can cover this logic as well.

Ok, I'll do it.